internal/rpc/core/events.go

```diff
@@ -126,3 +129,131 @@ func (env *Environment) UnsubscribeAll(ctx context.Context) (*coretypes.ResultUn
 	}
 	return &coretypes.ResultUnsubscribe{}, nil
 }
+
+// Events applies a query to the event log. If an event log is not enabled,
+// Events reports an error. Otherwise, it filters the current contents of the
+// log to return matching events.
+//
+// Events returns up to maxItems of the newest eligible event items. An item is
+// eligible if it is older than before (or before is zero), it is newer than
+// after (or after is zero), and its data matches the filter. A nil filter
+// matches all event data.
+//
+// If before is zero and no eligible event items are available, Events waits
+// for up to waitTime for a matching item to become available. The wait is
+// terminated early if ctx ends.
+//
+// If maxItems ≤ 0, a default positive number of events is chosen. The values
+// of maxItems and waitTime may be capped to sensible internal maxima without
+// reporting an error to the caller.
+func (env *Environment) Events(ctx context.Context,
+	filter *coretypes.EventFilter,
+	maxItems int,
+	before, after cursor.Cursor,
+	waitTime time.Duration,
+) (*coretypes.ResultEvents, error) {
+	if env.EventLog == nil {
+		return nil, errors.New("the event log is not enabled")
+	}
+
+	// Parse and validate parameters.
+	if maxItems <= 0 {
+		maxItems = 10
+	} else if maxItems > 100 {
+		maxItems = 100
+	}
+
+	const maxWaitTime = 30 * time.Second
+	if waitTime > maxWaitTime {
+		waitTime = maxWaitTime
+	}
+
+	query := tmquery.All
+	if filter != nil && filter.Query != "" {
+		q, err := tmquery.New(filter.Query)
+		if err != nil {
+			return nil, fmt.Errorf("invalid filter query: %w", err)
+		}
+		query = q
+	}
+
+	var info eventlog.Info
+	var items []*eventlog.Item
+	var err error
+	accept := func(itm *eventlog.Item) error {
+		// N.B. We accept up to one item more than requested, so we can tell how
+		// to set the "more" flag in the response.
+		if len(items) > maxItems {
+			return eventlog.ErrStopScan
+		}
+		if cursorInRange(itm.Cursor, before, after) && query.Matches(itm.Events) {
+			items = append(items, itm)
+		}
+		return nil
+	}
+
+	if waitTime > 0 && before.IsZero() {
+		ctx, cancel := context.WithTimeout(ctx, waitTime)
+		defer cancel()
+
+		// Long poll. The loop here is because new items may not match the query,
+		// and we want to keep waiting until we have relevant results (or time out).
+		cur := after
+		for len(items) == 0 {
+			info, err = env.EventLog.WaitScan(ctx, cur, accept)
```

Comment:
Q about this implementation: does WaitScan allow the full ctx to elapse or does it scan the list as soon as items are available? A client may have listed that they are willing to wait a long time for items and want a lot of items. If such a client arrives at an empty list that then has 1 item appended followed by a ton of items, I _think_ they will just get the 1 item back, which feels a little off since they are still willing to wait for the entire `waitTime` for their items.

Reply:
It returns as soon as it has something. That's how it's defined in the ADR but it wouldn't be hard to change it to keep trying. 

I think we probably want to return eagerly though: 
A client who's waiting has already found no matches, so they're right at the head of the log. Do they want to wait for more items or have the new data ASAP? We can't know, but if they want "more" they can easily call again (and probably reuse the same HTTP connection).

By contrast if they DIDN'T want to wait, they gave to kinda tune their timeout carefully and will wind up polling rapidly just to avoid stalling. 

So my intuition is that eager is better, but it isn't much work to change it if there's a good case for waiting.